components/codec/src/columnar/bytes.rs

```diff
@@ -12,37 +12,119 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use bytes_ext::{Buf, BufMut, Bytes};
-use snafu::ResultExt;
+use std::io::{Read, Write};
+
+use bytes_ext::{Buf, BufMut, Bytes, WriterOnBufMut};
+use lz4_flex::frame::{FrameDecoder as Lz4Decoder, FrameEncoder as Lz4Encoder};
+use snafu::{ensure, ResultExt};
 
 use crate::{
     columnar::{
+        Compress, DecodeContext, Decompress, InvalidCompression, InvalidVersion, NotEnoughBytes,
         Result, ValuesDecoder, ValuesDecoderImpl, ValuesEncoder, ValuesEncoderImpl, Varint,
     },
     varint,
 };
 
+/// The layout for the string/bytes:
+/// ```plaintext
+/// +-------------+--------------+------------+-----------------------+-----------------+
+/// | version(u8) | length_block | data_block | length_block_len(u32) | compression(u8) |
+/// +-------------+--------------+------------+-----------------------+-----------------+
+/// ```
+///
+/// Currently, the `compression` has two optional values:
+/// - 0: No compression over the data block
+/// - 1: the data block will be compressed if it is too long
+///
+/// And the lengths in the `length block` are encoded in varint.
+/// And the reason to put `length_block_len` and `compression` at the footer is
+/// for friendly decode.
+struct Encoding;
+
+impl Encoding {
+    const COMPRESSION_SIZE: usize = 1;
+    const LENGTH_BLOCK_LEN_SIZE: usize = 4;
+    const VERSION: u8 = 0;
+    const VERSION_SIZE: usize = 1;
+
+    fn decide_compression(data_block_len: usize, threshold: usize) -> Compression {
+        if data_block_len > threshold {
+            Compression::Lz4
+        } else {
+            Compression::NoCompression
+        }
+    }
+
+    fn decode_compression(v: u8) -> Result<Compression> {
+        let version = match v {
+            0 => Compression::NoCompression,
+            1 => Compression::Lz4,
+            _ => InvalidCompression { flag: v }.fail()?,
+        };
+
+        Ok(version)
+    }
+}
+
+/// The compression for [`Encoding`].
+///
+/// It is not allowed to be modified and only allowed to be appended with a new
+/// variant.
+#[derive(Clone, Copy, Default)]
+#[repr(C)]
+enum Compression {
+    #[default]
+    NoCompression = 0,
+    Lz4 = 1,
+}
+
 impl<'a> ValuesEncoder<&'a [u8]> for ValuesEncoderImpl {
+    /// The layout can be referred to the docs of [`Encoding`].
     fn encode<B, I>(&self, buf: &mut B, values: I) -> Result<()>
     where
         B: BufMut,
-        I: Iterator<Item = &'a [u8]>,
+        I: Iterator<Item = &'a [u8]> + Clone,
     {
-        for v in values {
-            debug_assert!(v.len() < u32::MAX as usize);
+        // Encode the `version`.
+        buf.put_u8(Encoding::VERSION);
+
+        // Encode the `length_block`.
+        let mut data_block_len = 0;
+        let mut length_block_len = 0;
+        for v in values.clone() {
```

Comment:
why clone is required here?

Reply:
The iterator will be consumed if not clone.